File: Kotlin/Euler021.kt
import java.lang.Math

fun main(args: Array<String>) {
    val target = 10000
    var properDivisorsSums: MutableMap<Int, Int> = mutableMapOf()
    var amicableSum = 0

    for (i in 1..(target-1)) {
        var divisorsSum = properDivisorsSums[i]
        if (divisorsSum == properDivisorsSums[i]) {
            divisorsSum = getProperDivisorsSum(i)
            properDivisorsSums[i] = divisorsSum
        }

        var otherDivisorsSum = properDivisorsSums[divisorsSum]
        if (otherDivisorsSum == null) {
            otherDivisorsSum = getProperDivisorsSum(divisorsSum as Int)
            properDivisorsSums[divisorsSum] = otherDivisorsSum
        }

        if (otherDivisorsSum == i && i != divisorsSum) {
            amicableSum += i
        }
    }

    println(amicableSum)
}

fun getProperDivisorsSum(n: Int): Int {
    var sum: Int = 1
    var lim: Int = Math.ceil(Math.sqrt(n.toDouble())).toInt()

    for (i: Int in 2..lim) {
        if (n % i == 0) {
            sum += i

            if (n != lim) {
                sum += n / i
            }
        }
    }

    return sum
}

File: Python/Euler021.py
#!/usr/bin/env python3
from itertools import chain
import math
import sys
if sys.version.startswith('2'):
    range = xrange


def main():
    target = 10000
    proper_divisors_sums = {}
    amicable_sum = 0

    for i in range(1, target):
        if i not in proper_divisors_sums:
            proper_divisors_sums[i] = get_proper_divisors_sum(i)
        divisors_sum = proper_divisors_sums[i]

        if divisors_sum not in proper_divisors_sums:
            proper_divisors_sums[divisors_sum] = get_proper_divisors_sum(divisors_sum)
        other_divisors_sum = proper_divisors_sums[divisors_sum]

        if other_divisors_sum == i and i != divisors_sum:
            amicable_sum += i

    print(amicable_sum)


def get_proper_divisors_sum(n):
    limit = math.ceil(math.sqrt(n)) + 1
    nums = ((i, n // i) for i in range(2, limit + 1) if n % i == 0)
    return sum(set(chain(*nums))) + 1


if __name__ == '__main__':
    main()
